Initial pcs property support

mbaldessari opened the pull request:
Add a function in order to support pcs properties natively.
Support is per cluster-wide properties or per-node properties.
Example:

pacemaker::property {"global-bar":
  property        => "bar",
  value           => "baz",
  force           => true,
  tries           => 1,
  try_sleep       => 1,
}

jguiditta commented:
I have to say that I am on the fence with this.  I like the idea to add this support, but on the other hand, I think this module should probably be trying to move away from execs, and instead use more puppet provider code.  That said, if nobody else strenuously objects, perhaps this could be a good interim step, and the impl under pacemaker::property could later change to call a provider.  Any thoughts on this @EmilienM (aside from 'this needs to get under jenkins and have more/better tests', which I agree with)?

EmilienM commented on README.md:
```diff
@@ -190,7 +196,16 @@ $clone_params succeeds as expected.
+      force           => true,
+      value           => "baz",
+      tries           => 1,
+      try_sleep       => 1,
```
please align the => correctly, too much spaces here.

EmilienM commented on README.md:
```diff
@@ -190,7 +196,16 @@ $clone_params succeeds as expected.
     # results in Debug: /usr/sbin/pcs resource create neutron-scale ocf:neutron:NeutronScale --clone globally-unique=true clone-max=3 interleave=true
-    
+
+    pacemaker::property {"global-bar":
```
just syntax:
```pacemaker::property { 'global-bar':```

EmilienM commented on manifests/property.pp:
```diff
@@ -0,0 +1,65 @@
+define pacemaker::property (
+  $property        = undef,
```
please align the = without useless spaces.

EmilienM commented on manifests/property.pp:
```diff
@@ -0,0 +1,65 @@
+  $tries           = 1,
+  $try_sleep       = 10,
+) {
+  if $property == undef {
```
if the value is required, I suggest to declare the param without default value. Puppet will fail if nothing is provided.

EmilienM commented on manifests/property.pp:
```diff
@@ -0,0 +1,65 @@
+  if $property == undef {
+    fail("Must provide property")
+  }
+  if ($ensure == present and $value == undef ) {
```
```if ($ensure == 'present') and ! $value {``` is a bit more puppetish. Undef is considered as a boolean in a conditional statement.

EmilienM commented on manifests/property.pp:
```diff
@@ -0,0 +1,65 @@
+  if $force {
+    $_force = "--force"
+  } else {
+    $_force = ""
```
That's ugly. I'm even not sure the ```pcs``` CLI allows to run "" in the arguments. I've seen many bugs in Puppet catalogs where people used to set ```""``` for empty values, and it failed.

```usr/sbin/pcs property set "" --node ...``` might causes issues.
You need to find another way IMHO, except if you tested it.

EmilienM commented on manifests/property.pp:
```diff
@@ -0,0 +1,65 @@
+
+  # Special-casing node branches due to https://bugzilla.redhat.com/show_bug.cgi?id=1302010
+  # (Basically pcs property show <property> will show all node properties anyway)
+  if $node != undef {
```
```if ! $node {``` is more clean.

EmilienM commented on tests/init.pp:
```diff
@@ -115,3 +115,12 @@
 }
+
+### Add a property
+pacemaker::property {"global-bar":
```
lint: ```pacemaker::property { 'global-bar':```

EmilienM commented:
Of course, unit tests are more than welcome.

mbaldessari commented:
Thanks for the feedback @EmilienM, much appreciated. I addressed all your comments (not sure why it is not showing as a new commit, but the existing one is definitely updated). I did test all scenarios before (with --force and without) but I agree with the risks of passing "" to pcs, so I amended that as
well.

Regarding adding this to CI and unit tests, I am happy to help but might need a kick or two in the right direction.

EmilienM commented on manifests/stonith.pp:
```diff
@@ -1,19 +1,13 @@
-        require => [ Exec["wait-for-settle"],
-                     Class['::pacemaker::corosync']
-                   ],
+    pacemaker::property { "Disable STONITH":
```
lint: use ' instead of ".
Use " only if you have variables.

EmilienM commented on manifests/stonith.pp:
```diff
@@ -1,19 +1,13 @@
-        require => [ Exec["wait-for-settle"],
-                     Class['::pacemaker::corosync']
-                   ],
+    pacemaker::property { "Enable STONITH":
```
dito

EmilienM commented:
@mbaldessari a last minor comment and code looks good to me.
As a note for later, we might want to setup some acceptance tests (beaker) and unit tests for this modules.
Merging code without any tests really make me sick.

mbaldessari commented:
Thanks @EmilienM . I have amended the commit and fixed a couple of other puppet-lint complaints.

Fully agreed on the unit tests. I will tackle them in a separate pull request. If you have a puppet module that you consider a good example regarding unit testing, I am all ears. Otherwise I'll just read up on the topic and wade my way through ;)

EmilienM commented:
I'm blindly merging this patch that won't hopefully not break anything.
Thanks for your contribution and your patience, very appreciated here.
